src/adapters/webpush/mod.rs

```diff
@@ -0,0 +1,359 @@
+            .body(Body::BufBody(&enc.output, enc.output.len()))
+            .send() {
+                Ok(x) => x,
+                Err(e) => { warn!("notify subscription {} failed: {:?}", self.push_uri, e); return; }
+            };
+
+        info!("notified subscription {} (status {:?})", self.push_uri, res.status);
+    }
+}
+
+pub struct WebPush<C> {
+    controller: C,
+    getter_resource_id: Id<Getter>,
+    getter_subscription_id: Id<Getter>,
+    setter_resource_id: Id<Setter>,
+    setter_subscribe_id: Id<Setter>,
+    setter_unsubscribe_id: Id<Setter>,
+    setter_notify_id: Id<Setter>,
+}
+
+impl<C: Controller> WebPush<C> {
+    pub fn id() -> Id<AdapterId> {
+        Id::new("webpush@link.mozilla.org")
+    }
+
+    pub fn service_webpush_id() -> Id<ServiceId> {
+        Id::new("service:webpush@link.mozilla.org")
+    }
+
+    pub fn getter_resource_id() -> Id<Getter> {
+        Id::new("getter:resource.webpush@link.mozilla.org")
+    }
+
+    pub fn getter_subscription_id() -> Id<Getter> {
+        Id::new("getter:subscription.webpush@link.mozilla.org")
+    }
+
+    pub fn setter_resource_id() -> Id<Setter> {
+        Id::new("setter:resource.webpush@link.mozilla.org")
+    }
+
+    pub fn setter_subscribe_id() -> Id<Setter> {
+        Id::new("setter:subscribe.webpush@link.mozilla.org")
+    }
+
+    pub fn setter_unsubscribe_id() -> Id<Setter> {
+        Id::new("setter:unsubscribe.webpush@link.mozilla.org")
+    }
+
+    pub fn setter_notify_id() -> Id<Setter> {
+        Id::new("setter:notify.webpush@link.mozilla.org")
+    }
+}
+
+impl<C: Controller> Adapter for WebPush<C> {
+    fn id(&self) -> Id<AdapterId> {
+        Self::id()
+    }
+
+    fn name(&self) -> &str {
+        ADAPTER_NAME
+    }
+
+    fn vendor(&self) -> &str {
+        ADAPTER_VENDOR
+    }
+
+    fn version(&self) -> &[u32;4] {
+        &ADAPTER_VERSION
+    }
+
+    fn fetch_values(&self, mut set: Vec<Id<Getter>>) -> ResultMap<Id<Getter>, Option<Value>, Error> {
+        set.drain(..).map(|id| {
+            let user_id = 1; // FIXME: currently logged in user
+
+            macro_rules! getter_api {
+                ($getter:ident, $getter_id:ident, $getter_type:ident) => (
+                    if id == self.$getter_id {
+                        match self.$getter(user_id) {
+                            Ok(data) => {
+                                let rsp = $getter_type::new(data);
+                                return (id, Ok(Some(Value::Json(Arc::new(Json(serde_json::to_value(&rsp)))))));
+                            },
+                            Err(_) => return (id, Err(Error::InternalError(InternalError::InvalidInitialService)))
+                        };
+                    }
+                ) 
+            }
+
+            getter_api!(user_subscriptions, getter_subscription_id, SubscriptionGetter);
+            getter_api!(resources, getter_resource_id, ResourceGetter);
```

Comment:
that may be a bit clearer to name these get_user_subscriptions and get_resources